### PR TITLE
Redirect prev

### PR DESCRIPTION
**Pull request by GuillaumeCisco**
Fix issue exposed in https://github.com/faceyspacey/redux-first-router/issues/65#issuecomment-320516244

### CODE REVIEW

**Comment by faceyspacey** [comment]
better place to put it...and this solved your need? partially or perfectly solved?

**Comment by faceyspacey** [comment]
...i can see the tests failing exactly as they should. excellent job. just run `yarn test -- -u` and then push the updated snapshots.

**Comment by GuillaumeCisco** [comment]
It totally solves my need.
Happy tests fail correctly with this modification, I just pushed the snapshots.

**Comment by faceyspacey** [comment]
I'm not sure why more tests are failing now. It seems more snapshots were changed based on how things ran on your computer.

If possible revert all the snapshots except for the 2 initially failing tests. One sure-fire way to make it pass is to manually change the `prev` property in those 2 initially failing tests. And then make sure know other snapshot files are changed.

**Comment by GuillaumeCisco** [comment]
I've just repushed the connectedRoutes snapshot after updating tests
I now only have problem with QUERY `tests`. I don't know if it's normal or not.
Can you help me on this @faceyspacey ?

**Comment by faceyspacey** [comment]
set hasSSR back to undefined.

for some reason, your jest isn't setting the window object:

https://github.com/faceyspacey/redux-first-router/blob/60f046ff214710fc412469347d068ec0928f4c7e/src/pure-utils/isServer.js

`window` is supposed to be defined while running Jest. What OS are you on?

Either way, just make the query tests have the same `hasSSR` value as before, which is `undefined`.

**Comment by faceyspacey** [comment]
also you deleted 2 tests. they can't be deleted. the `prev` key just needs to be changed to the correct location.

File: __tests__/__snapshots__/connectRoutes.js.snap [hunk 1]
```diff
@@ -295,48 +295,6 @@ Object {
 }
 `;
 
-exports[`middleware if onBeforeChange dispatches redirect, route changes with kind === "redirect" 1`] = `
```
Comment (faceyspacey) [inline __tests__/__snapshots__/connectRoutes.js.snap]: these cant be deleted. just change the `prev` key to the new object

File: __tests__/__snapshots__/connectRoutes.js.snap [hunk 2]
```diff
@@ -71,7 +71,7 @@ Object {
 exports[`QUERY: dispatched as action.payload.query 1`] = `
 Object {
   "location": Object {
-    "hasSSR": undefined,
+    "hasSSR": true,
```
Comment (faceyspacey) [inline __tests__/__snapshots__/connectRoutes.js.snap]: these all need to stay the previous value.

File: __tests__/connectRoutes.js [hunk 3]
```diff
@@ -114,7 +114,13 @@ describe('middleware', () => {
     expect(location.kind).toEqual('redirect')
     expect(location.type).toEqual('THIRD')
     expect(history.entries.length).toEqual(2)
-    expect(location).toMatchSnapshot()
+    expect(location.prev).toEqual({
```
Comment (faceyspacey) [inline __tests__/connectRoutes.js]: this still needs to be a snapshot, not `toEqual`. update the snapshot just for this test so u dont break other ones:

```js
yarn test --testPathPattern regex-for-test-name

# after it works, replace the snapshots with
yarn test --testPathPattern regex-for-test-name -- -u
```

Comment (faceyspacey) [inline __tests__/connectRoutes.js]: ok remove all test related code back to your original PR. Then make sure on the "conversation" tab I'm added in the top right with the ability to edit your PR.

**Comment by faceyspacey** [comment]
solid. Thanks for this. It will be on NPM in a few minutes after the next build automatically pushes it there.

**Comment by GuillaumeCisco** [comment]
THank you @faceyspacey 
Still don't know why my computer settings prevent me from running correctly the tests

**Comment by faceyspacey** [comment]
all good. ...log the value of `window` in tests sometime and see if it's defined. then google why Jest's window object may not be defined.

**Comment by GuillaumeCisco** [comment]
Just log it, window is always defined but sometimes `!!window.SSRtest` is ever false or true.
I'm not sure it comes from an undefined window object.

**Comment by faceyspacey** [comment]
Jest has a param to make tests run sync. Try that. Perhaps multiple async tests are overlapping.

**Comment by GuillaumeCisco** [comment]
Just tried with the `--runInBand` with `yarn test -- --runInBand` command. But same thing.
Still a mistery for me.

**Comment by GuillaumeCisco** [comment]
Something interesting, if I run the failing tests individually, it works.
Looks like running the tests one after another is not like running them isolated, maybe something is overlapping yes...

**Comment by GuillaumeCisco** [comment]
Just changed all the `delete global.window.SSRtest` by `global.window.SSRtest = false` and the error disappeared. Clearly don't know why the `delete` one does not work.
Also, I don't think the isServer method should depend on a variable only created for testing purpose. Can't we mock this method during the time of the test for returning true ?

I know have other issues with 
nestHistory formats simplified history object for action + state
```
Object {
   "entries": Array [
     Object {
       "hash": "",
-      "key": undefined,
+      "key": "345678",
       "pathname": "/",
       "search": "",
       "state": undefined,
     },
```
createLocationReducer() - maintains address bar pathname state and current + previous matched location-aware actions
```
"hasSSR": undefined,
   "history": Object {
     "entries": Array [
       Object {
         "hash": "",
-        "key": undefined,
+        "key": "345678",
         "pathname": "/first",
         "search": "",
         "state": undefined,
       },
```
scrolls to top on route change when options.scrollTop === true
```
Error: expect(jest.fn()).toHaveBeenCalled()

Expected mock function to have been called.
```
Maybe it will help in future developement

**Comment by faceyspacey** [comment]
u can mock it, but then you have to create a whole other test file just for that tests. mocks are file-wide. i wish they were dynamic and can be changed per test. i prefer not to have a million test files, but one test file per original module file.

...the key is from history package. i don't know why it's appearing right now--probably because the browser history is being used instead of the memory history due to the IS_TEST variable being different.

I wouldn't worry about it for now. It just seems Jest works differently on Linux.

**Pull request merged**